ddl/job_table.go

```diff
@@ -194,28 +165,45 @@ func hasSysDB(job *model.Job) bool {
 	return false
 }
 
-func (d *ddl) handleUpgradingState(se *sess.Session, job *model.Job) error {
-	if !d.stateSyncer.IsUpgradingState() {
-		if !job.IsPausedBySystem() || hasSysDB(job) {
-			return nil
+func (d *ddl) processJobDuringUpgrade(sess *sess.Session, job *model.Job) (isRunnable bool, err error) {
+	if d.stateSyncer.IsUpgradingState() {
+		if hasSysDB(job) {
+			return true, nil
 		}
-		_, err := ResumeJobsBySystem(se.Session(), []int64{job.ID})
-		if err != nil {
-			logutil.BgLogger().Warn("[ddl] resume user DDL by system failed", zap.Stringer("job", job), zap.Error(err))
-			return err
+		var errs []error
+		// During binary upgrade, pause all running DDL jobs
+		errs, err = PauseJobsBySystem(sess.Session(), []int64{job.ID})
+		if errs != nil && errs[0] != nil {
+			if dbterror.ErrPausedDDLJob.Equal(errs[0]) {
+				// In case of 'Pausing', we need to deliver the job to DDL
+				// worker to pause the background reorganization workers.
+				return true, nil
+			}
+			// During upgrade, there are jobs that may not be paused because of
+			// some serious error. Then, we would not run it and just return
+			// unless the upgrade finished.
+			return false, errs[0]
 		}
-		logutil.BgLogger().Info("[ddl] resume user DDL by system successful", zap.Stringer("job", job))
-		return nil
-	}
-	if job.IsPausing() {
-		return nil
+
+		// The job has been issued the command 'pause', which has been turned
+		// to be 'pausing' in job table, then we need to deliver it to ddl
+		// worker to turn it to be 'paused'
+		job.State = model.JobStatePausing
```

Comment:
Could we remove it?

Reply:
Sure, it just is faster to continue the running this time.